Fix #2132

adrianbroher opened the pull request:
Encode the the DEFLATE compressed section of a save game as base64

This prevents the corruption during loading when linking against
boost::serialization 1.66.0 or later.

Fixes #2132

adrianbroher committed:
Add boost::iterators::base64_filter implementation

This header implements a boost::iterators base64_filter.  The
implementation is licensed under the Boost Software License, Version 1.0

Upstream released this file as two separated files at
http://hamigaki.osdn.jp/doc/html/iostreams.html.

* hamigaki/iostreams/filter/base64.hpp
* hamigaki/iostreams/arbitrary_positional_facade.hpp

The implementation was moved into the `boost::iterators` namespace and
both files were merged into a single one.

adrianbroher committed:
Encode the the DEFLATE compressed section of a save game as base64

This prevents the corruption during loading when linking against
boost::serialization 1.66.0 or later.

Fixes #2132

adrianbroher committed:
Call destructor of serialization archive for DEFLATE save game section

Previous implementations of compressed save games didn't wrote the
closing XML tag reliably. By invoking the destructor of the
serialization archive of the DEFLATE save game section the closing tag
is now written reliable for all known boost versions.

Vezzra reviewed (approved):
Fixes the issue on OSX (for saves created with this PR of course).

geoffthemedio committed:
Raise std::invalid_argument when loading zlib-xml savegame with boost 1.66.0+

Starting with version 1.66.0 boost::serialization doesn't load the old
zlib-xml savegames anymore.  When a user now saves the game as
compressed xml the zb64-xml format is used.  To avoid crashing on the
user the server doesn't read the old savegame format anymore.
